misc/python/materialize/parallel_workload/action.py

```diff
@@ -0,0 +1,137 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import random
+from typing import List
+
+import pg8000
+
+from materialize.parallel_workload.database import Database, DBObject, Table
+from materialize.parallel_workload.execute import execute
+
+MAX_TABLES = 25
+MAX_VIEWS = 25
+
+
+class Action:
+    rng: random.Random
+    database: Database
+
+    def __init__(self, rng: random.Random, database: Database):
+        self.rng = rng
+        self.database = database
+
+    def run(self, cur: pg8000.Cursor) -> None:
+        raise NotImplementedError
+
+
+class SelectAction(Action):
+    def run(self, cur: pg8000.Cursor) -> None:
+        tables_views: List[DBObject] = [*self.database.tables, *self.database.views]
+        table = self.rng.choice(tables_views)
+        # TODO: More interesting expressions
```

Comment:
An expression would not be very relevant for a DDL-focused test. Consider just using `*` here.

What would be quite relevant though is have multiple objects per `SELECT`. So , with some probability, select 2 database objects and join them, with a restrictive join clause and a `LIMIT 1` to avoid too large result sets.

Reply:
I noticed that with `LIMIT 1` many queries are still getting stuck in the join-selects. statement_timeout was also not helpful. So I'll limit the number of views which are joined together and add DeleteActions so that we don't accumulate too much data.

Reply:
I think for the purpose of fishing for DDL issues, a max of two objects per view would be sufficient. An alternative to LIMIT 1 would be to provide an impossible WHERE condition (but please run an EXPLAIN manually to confirm that the query still has the join and has not been optimized to 1-table or to no tables). After all, this framework does not validate the output of the SELECTs, so the only thing that matters is that the query is not `Constant` an actual dataflow that references more than one database object is constructed .